web/src/message_fetch.ts

```diff
@@ -313,26 +314,22 @@ export function load_messages(opts: MessageFetchOptions, attempt = 1): void {
     // But support for the all_messages_data sharing of data with
     // the combined feed view and the (hacky) page_params.narrow feature
     // requires a somewhat ugly bundle of conditionals.
-    let narrow_data_string = "";
-    if (msg_list_data.filter.is_in_home()) {
-        if (page_params.narrow_stream !== undefined) {
-            narrow_data_string = JSON.stringify(page_params.narrow);
-        }
-        // Otherwise, we don't pass narrow for the combined feed view; this is
-        // required to display messages if their muted status changes without a new
-        // network request, and so we need the server to send us message history from muted
-        // streams and topics even though the combined feed view's in:home
-        // operators will filter those.
-    } else {
-        let terms = msg_list_data.filter.public_terms();
-        if (page_params.narrow !== undefined) {
-            terms = [...terms, ...page_params.narrow];
-        }
-        narrow_data_string = JSON.stringify(terms);
+    let narrow_data: NarrowTerm[] = [];
+    if (page_params.narrow !== undefined) {
+        narrow_data = page_params.narrow;
     }
-
-    if (narrow_data_string !== "") {
-        data.narrow = handle_operators_supporting_id_based_api(narrow_data_string);
+    if (!msg_list_data.filter.is_in_home()) {
+        const message_list_terms = msg_list_data.filter.public_terms();
+        narrow_data = [...narrow_data, ...message_list_terms];
```

Comment:
It looks like the old code had the public terms first and then the `page_params` terms, and here you have them in the opposite order (public terms second).

Reply:
I wasn't thinking about that order. Updated so that the narrow is built with the filter's public terms first and then the page params narrow data and finally the spectator narrow term.